### ISSUE TITLE
Fix blockchain hanging forever in case code throws between a semaphore lock/unlock

### ISSUE THREAD

**Issue by jochem-brouwer** [ethereumjs/ethereumjs-vm#877]
In [blockchain/index.ts](https://github.com/ethereumjs/ethereumjs-vm/blob/d246b2b2eea3f518eb04490325df37f5491bdf20/packages/blockchain/src/index.ts#L304) we have several blocks of code which use the semaphore lock/unlock feature. The problem here is that in case the code between lock/unlock throws, the lock is never released and thus any call to `wait` will hang forever.

We should probably add something like `safeLock` which runs a block of code, but in case this block of code throws, then unlock the lock and re-raise the error.

It is also worth looking into any other package which uses semaphore: these kinds of bugs would be rather hard to reproduce, but they would render the package useless in case you run into them.

**Comment by jochem-brouwer** [comment]
Only blockchain uses `semaphore` and the fix is rather small so I will put this in #849

**Comment by holgerd77** [comment]
Makes sense, this would be something like...

```typescript
await this._lock.wait()
try {
  await this._putBlockOrHeader(block, isGenesis)
} catch (err) {
  this._lock.release()
  throw(err)
}

..I would assume?

**Comment by jochem-brouwer** [comment]
Exactly 👍

**Comment by jochem-brouwer** [comment]
Closed by #849
